src/python/pants/engine/rules.py

```diff
@@ -39,33 +39,136 @@ def side_effecting(cls):
-        return isinstance(node, ast.Name) and node.id == Get.__name__
-
-    def _is_get(self, node: ast.AST) -> bool:
-        """Check if the node looks like a Get(...) or Get[X](...) call."""
-        if isinstance(node, ast.Call):
-            if self._matches_get_name(node.func):
-                return True
-            if isinstance(node.func, ast.Subscript) and self._matches_get_name(node.func.value):
-                return True
-            return False
-        return False
-
-    def visit_Call(self, node: ast.Call) -> None:
-        if self._is_get(node):
-            self._gets.append(Get.extract_constraints(node))
+    @frozen_after_init
+    @dataclass(unsafe_hash=True)
+    class _GetDescriptor:
+        product_type_name: str
+        subject_arg_exprs: Tuple[ast.expr, ...]
+
+        def __init__(
+            self, product_type_expr: ast.expr, subject_arg_exprs: Iterable[ast.expr]
+        ) -> None:
+            if not isinstance(product_type_expr, ast.Name):
+                raise ValueError(
+                    f"Unrecognized type argument T for Get[T]: " f"{ast.dump(product_type_expr)}"
+                )
+            self.product_type_name = product_type_expr.id
+            self.subject_arg_exprs = tuple(subject_arg_exprs)
+
+    def _identify_source(self, node: Union[ast.expr, ast.stmt]) -> str:
+        start_pos = f"{node.lineno}:{node.col_offset}"
+
+        end_lineno, end_col_offset = [
+            getattr(node, attr, None) for attr in ("end_lineno", "end_col_offset")
+        ]
+        end_pos = f"-{end_lineno}:{end_col_offset}" if end_lineno and end_col_offset else ""
+
+        return f"{self._source_file} at {start_pos}{end_pos}"
+
+    def _extract_get_descriptor(self, call_node: ast.Call) -> Optional[_GetDescriptor]:
+        """Check if the node looks like a Get[T](...) call."""
+        if not isinstance(call_node.func, ast.Subscript):
+            if not isinstance(call_node.func, ast.Name):
+                return None
+            if call_node.func.id != "Get":
+                return None
+            return self._GetDescriptor(
+                product_type_expr=call_node.args[0], subject_arg_exprs=call_node.args[1:]
+            )
+
+        subscript_func = call_node.func
+        if not isinstance(subscript_func.slice, ast.Index):
+            return None
+        node_name = subscript_func.value
+        if not isinstance(node_name, ast.Name):
+            return None
+        if node_name.id != "Get":
+            return None
+
+        get_descriptor = self._GetDescriptor(
+            product_type_expr=subscript_func.slice.value, subject_arg_exprs=call_node.args
+        )
+
+        # TODO(John Sirois): Turn this on and update Pants own codebase to not trigger the warning
+        #  in a follow-up.
+        #  https://github.com/pantsbuild/pants/issues/9899
+        deprecated_conditional(
+            predicate=lambda: False,
+            removal_version="1.31.0.dev0",
```

Comment:
Changed.

Reply:
Pardon me. I gave you bad advice. If we're trying to land this refactor in 1.29.0rc0 today—which I think is a good idea—then turning the deprecations off won't make it in time for 1.29. Meaning we deprecate in 1.30 and remove in 1.31.

But, because this deprecation never triggers in 1.29, we can forward fix this.